### PR TITLE
[release 3-11] Bug 1753593: [Reg] openshift_logging_elasticsearch_kibana_index_mode…

### PR DESCRIPTION
**Pull request by nhosoi**
… cannot be changed by inventory variable

1) Module logging_patch creates a patch to modify the new file to the original.
2) The configmap to be patched in patch_configmap_file.yaml is the new file.

### CODE REVIEW

**Comment by richm** [comment]
> I think there is some overlap with #11917
> 
> Can we sort out which approach we will take for the elasticsearch config?
> Personally, I would rather we stopped trying to preserve custom changes via patches

I think the root cause of https://github.com/openshift/openshift-ansible/pull/11917 is that the config is too old.  I think this PR is unrelated.  @nhosoi why is openshift_logging_elasticsearch_kibana_index_mode not able to be changed in the ansible inventory, and how does this PR fix that problem?

**Comment by nhosoi** [comment]
> > I think there is some overlap with #11917
> > Can we sort out which approach we will take for the elasticsearch config?
> > Personally, I would rather we stopped trying to preserve custom changes via patches
> 
> I think the root cause of #11917 is that the config is too old. I think this PR is unrelated. @nhosoi why is openshift_logging_elasticsearch_kibana_index_mode not able to be changed in the ansible inventory, and how does this PR fix that problem?

Well, 

> > I think there is some overlap with #11917
> > Can we sort out which approach we will take for the elasticsearch config?
> > Personally, I would rather we stopped trying to preserve custom changes via patches
> 
> I think the root cause of #11917 is that the config is too old. I think this PR is unrelated. @nhosoi why is openshift_logging_elasticsearch_kibana_index_mode not able to be changed in the ansible inventory, and how does this PR fix that problem?

The logging_patch module [1] generates a patch, which looks to me the patch is for modifying a new configmap to an original one.  (I can attach the generated patch to this PR in several hours as an example, which is in my another laptop...)  This PR is making the direction opposite.

[1] https://github.com/openshift/openshift-ansible/blob/release-3.11/roles/openshift_logging/library/logging_patch.py

**Comment by nhosoi** [comment]
As described in https://bugzilla.redhat.com/show_bug.cgi?id=1753593#c0, the initial value of openshift.kibana.index.mode is unique.  Then,  they reran ansible-playbook with "openshift.kibana.index.mode=shared_ops" set in the inventry file to replace the value unique with shared_ops.  But openshift.kibana.index.mode wound not get updated to shared_ops.

[1] is the patch returned from roles/openshift_logging/library/logging_patch.py.  It does not look to me it is changing openshift.kibana.index.mode value from unique to shared_ops, but doing an opposite...

[1]
```
--- 
+++ 
@@ -1,69 +1,57 @@
 cluster:
   name: ${CLUSTER_NAME}
-
+discovery.zen:
+  minimum_master_nodes: ${NODE_QUORUM}
+  ping.unicast.hosts: logging-es-ops-cluster
+gateway:
+  expected_nodes: ${RECOVER_EXPECTED_NODES}
+  recover_after_nodes: ${NODE_QUORUM}
+  recover_after_time: ${RECOVER_AFTER_TIME}
+io.fabric8.elasticsearch.kibana.mapping.app: /usr/share/elasticsearch/index_patterns/com.redhat.viaq-openshift.index-pattern.json
+io.fabric8.elasticsearch.kibana.mapping.empty: /usr/share/elasticsearch/index_patterns/com.redhat.viaq-openshift.index-pattern.json
+io.fabric8.elasticsearch.kibana.mapping.ops: /usr/share/elasticsearch/index_patterns/com.redhat.viaq-openshift.index-pattern.json
+network:
+  host: 0.0.0.0
+node:
+  data: ${HAS_DATA}
+  master: ${IS_MASTER}
+  max_local_storage_nodes: 1
+  name: ${DC_NAME}
+openshift.config:
+  project_index_prefix: project
+  time_field_name: '@timestamp'
+  use_common_data_model: true
+openshift.kibana.index.mode: unique
+openshift.operations.allow_cluster_reader: false
+openshift.searchguard:
+  keystore.path: /etc/elasticsearch/secret/admin.jks
+  truststore.path: /etc/elasticsearch/secret/searchguard.truststore
+path:
+  data: /elasticsearch/persistent/${CLUSTER_NAME}/data
+  logs: /elasticsearch/persistent/${CLUSTER_NAME}/logs
 script:
   inline: true
   stored: true
-
-node:
-  name: ${DC_NAME}
-  master: ${IS_MASTER}
-  data: ${HAS_DATA}
-  max_local_storage_nodes: 1
-
-network:
-  host: 0.0.0.0
-
-discovery.zen:
-  ping.unicast.hosts: logging-es-ops-cluster
-  minimum_master_nodes: ${NODE_QUORUM}
-
-gateway:
-  recover_after_nodes: ${NODE_QUORUM}
-  expected_nodes: ${RECOVER_EXPECTED_NODES}
-  recover_after_time: ${RECOVER_AFTER_TIME}
-
-io.fabric8.elasticsearch.kibana.mapping.app: /usr/share/elasticsearch/index_patterns/com.redhat.viaq-openshift.index-pattern.json
-io.fabric8.elasticsearch.kibana.mapping.ops: /usr/share/elasticsearch/index_patterns/com.redhat.viaq-openshift.index-pattern.json
-io.fabric8.elasticsearch.kibana.mapping.empty: /usr/share/elasticsearch/index_patterns/com.redhat.viaq-openshift.index-pattern.json
-
-openshift.config:
-  use_common_data_model: true
-  project_index_prefix: "project"
-  time_field_name: "@timestamp"
-
-openshift.searchguard:
-  keystore.path: /etc/elasticsearch/secret/admin.jks
-  truststore.path: /etc/elasticsearch/secret/searchguard.truststore
-
-openshift.operations.allow_cluster_reader: false
-
-openshift.kibana.index.mode: shared_ops
-
-path:
-  data: /elasticsearch/persistent/${CLUSTER_NAME}/data
-  logs: /elasticsearch/persistent/${CLUSTER_NAME}/logs
-
 searchguard:
   authcz.admin_dn:
   - CN=system.admin,OU=OpenShift,O=Logging
-  config_index_name: ".searchguard"
+  config_index_name: .searchguard
   ssl:
+    http:
+      clientauth_mode: OPTIONAL
+      enabled: true
+      keystore_filepath: /etc/elasticsearch/secret/key
+      keystore_password: kspass
+      keystore_type: JKS
+      truststore_filepath: /etc/elasticsearch/secret/truststore
+      truststore_password: tspass
+      truststore_type: JKS
     transport:
       enabled: true
       enforce_hostname_verification: false
-      keystore_type: JKS
       keystore_filepath: /etc/elasticsearch/secret/searchguard.key
       keystore_password: kspass
-      truststore_type: JKS
+      keystore_type: JKS
       truststore_filepath: /etc/elasticsearch/secret/searchguard.truststore
       truststore_password: tspass
-    http:
-      enabled: true
-      keystore_type: JKS
-      keystore_filepath: /etc/elasticsearch/secret/key
-      keystore_password: kspass
-      clientauth_mode: OPTIONAL
       truststore_type: JKS
-      truststore_filepath: /etc/elasticsearch/secret/truststore
-      truststore_password: tspass
```

**Comment by richm** [comment]
Why is the diff so large?  Was the elasticsearch.yml template reformatted?  That is, it doesn't look like the actual contents changed, it looks like they have been moved around, like being fed through a yaml parser, then output by a yaml formatter . . .

**Comment by nhosoi** [comment]
> Why is the diff so large? Was the elasticsearch.yml template reformatted? That is, it doesn't look like the actual contents changed, it looks like they have been moved around, like being fed through a yaml parser, then output by a yaml formatter . . .

The diff is just a dump returned from logging_patch.py, which generates a diff between current.yml and elasticsearch.yml.  I have both files and indeed the difference is that large.  Not sure about reformatting, though...  

Please note that current.yml has the new value "shared_ops" and elsticsearch.yml has the old one as follows...
```
$ grep kibana.index.mode *.yml
current.yml:openshift.kibana.index.mode: shared_ops
elasticsearch.yml:openshift.kibana.index.mode: unique
```

**Comment by richm** [comment]
Yes, the diff is indeed changing
```
-openshift.kibana.index.mode: shared_ops
```
into
```
+openshift.kibana.index.mode: unique
```
So yes, it looks like the fix here will be to reverse the order.

But the bigger problem is the huge diff - that means, the code will check for a diff, and will _always_ find a diff, even if *there is really no change*.

**Comment by ewolinetz** [comment]
Another concern i have here though is that reversing the order means that we'd lose custom settings the users may have provided. This goes back to me wondering about the overlap with #11917 and whether or not we would just opt to no longer try to patch in prior changes.

**Comment by jcantrill** [comment]
> Another concern i have here though is that reversing the order means that we'd lose custom settings the users may have provided. This goes back to me wondering about the overlap with #11917 and whether or not we would just opt to no longer try to patch in prior changes.

The more we embrace the operator model where we do something or you the admin take the control the better.

**Comment by nhosoi** [comment]
> Another concern i have here though is that reversing the order means that we'd lose custom settings the users may have provided. This goes back to me wondering about the overlap with #11917 and whether or not we would just opt to no longer try to patch in prior changes.

I think it's safer that this PR should wait for #11917 merged and reinvestigate...

**Comment by richm** [comment]
> I think it's safer that this PR should wait for #11917 merged and reinvestigate...

Then can someone review and `/lgtm` https://github.com/openshift/openshift-ansible/pull/11917 ?

**Comment by nhosoi** [comment]
Observations:
According to the python doc [1], the order of the args is supposed to be (original, new) as in this patch.
[1] https://docs.python.org/3/library/difflib.html#difflib.unified_diff
[2] https://github.com/openshift/openshift-ansible/pull/11918/files#diff-4f610dcaf03a5a5eb5992801c0bac625R112-R113

The reason why the diff is this large is likely we compare the new with the original at [3].  The new file is vanilla, just created from the template elasticsearch.yml.j2 at [4].  Then, it's combined with override_config and passed to to_nice_yaml(indent=2) after the comparison.  That is, the elasticsearch configmap always has the diff and would be _modified_ since the new file is vanilla and the existing one is treated.  So, I think that part is as expected?  (Please note that other pods do not do the treatment.  It looks just elasticsearch does.)

[3] https://github.com/openshift/openshift-ansible/blob/release-3.11/roles/openshift_logging_elasticsearch/tasks/main.yaml#L287-L289
[4] https://github.com/openshift/openshift-ansible/blob/release-3.11/roles/openshift_logging_elasticsearch/tasks/main.yaml#L224-L226
[5] https://github.com/openshift/openshift-ansible/blob/release-3.11/roles/openshift_logging_elasticsearch/tasks/main.yaml#L287-L289

In conclusion, I think this PR is reasonable. :)

**Comment by richm** [comment]
hmm - looks like it might be related to https://github.com/openshift/aos-cd-jobs/pull/1985
```
+ scp -F /var/lib/jenkins/jobs/test_pull_request_openshift_ansible_logging-release-3.11/workspace/.config/origin-ci-tool/inventory/.ssh_config /var/lib/jenkins/.config/gcloud/gcs-publisher-credentials.json openshiftdevel:/data/credentials.json
Permission denied (publickey,gssapi-keyex,gssapi-with-mic).
lost connection
...
+ ssh -F /var/lib/jenkins/jobs/test_pull_request_openshift_ansible_logging-release-3.11/workspace/.config/origin-ci-tool/inventory/.ssh_config openshiftdevel sudo chmod o+rw /etc/environment
Permission denied (publickey,gssapi-keyex,gssapi-with-mic).
++ export status=FAILURE
++ status=FAILURE
+ set +o xtrace
########## FINISHED STAGE: FAILURE: FORWARD PARAMETERS TO THE REMOTE HOST [00h 00m 00s] ##########
```

but what's really strange is that this succeeds?
```
TASK [aws-up : wait for SSH to be available] ***********************************
task path: /var/lib/jenkins/origin-ci-tool/e7531a43acf671d647d0c19b3d1fddbc79373006/lib/python2.7/site-packages/oct/ansible/oct/playbooks/provision/roles/aws-up/tasks/main.yml:175
ok: [localhost] => {
    "changed": false, 
    "elapsed": 64, 
    "generated_timestamp": "2019-10-08 21:01:20.916634", 
    "path": null, 
    "port": 22, 
    "search_regex": null, 
    "state": "started"
}
```
@dobbymoodge @stevekuznetsov any ideas?

**Comment by richm** [comment]
don't bother to retest - 3.11 logging CI is broken because of https://bugzilla.redhat.com/show_bug.cgi?id=1766619

**Comment by richm** [comment]
aws flake
/retest

**Comment by nhosoi** [comment]
Wow...  The logging test finally passed!?!?!?  Thank you sooooo much, @richm!!

**Comment by richm** [comment]
> Wow... The logging test finally passed!?!?!? Thank you sooooo much, @richm!!

Well, not exactly - I logged into the CI machine and manually applied the patch from https://github.com/openshift/aos-cd-jobs/pull/2040 so that I could test that the patch worked

**Comment by nhosoi** [comment]
> > Wow... The logging test finally passed!?!?!? Thank you sooooo much, @richm!!
> 
> Well, not exactly - I logged into the CI machine and manually applied the patch from [openshift/aos-cd-jobs#2040](https://github.com/openshift/aos-cd-jobs/pull/2040) so that I could test that the patch worked

I see.  But great news that now we have a working patch!

**Comment by nhosoi** [comment]
@richm, do you think we should rerun the CI tests since openshift/aos-cd-jobs#2040 was merged?    Also, could you please review the PR one more time?  Thanks.

**Comment by richm** [comment]
> @richm, do you think we should rerun the CI tests since [openshift/aos-cd-jobs#2040](https://github.com/openshift/aos-cd-jobs/pull/2040) was merged? Also, could you please review the PR one more time? Thanks.

I think @ewolinetz or @jcantrill should review and approve
